libc/functions/i/ioctl.part-impl.md

```diff
@@ -0,0 +1,77 @@
+# Synopsis 
+`#include <ioctl.h>`</br>
+
+` int ioctl(int fildes, int request, ... /* arg */);`</br>
+
+## Status
+Partially implemented
+## Conformance
+IEEE Std 1003.1-2017
+## Description
+
+
+The `ioctl()` function shall perform a variety of control functions on devices. For fildes not referring to devices, the functions performed by this call are unspecified.
+
+The fildes argument is an open file descriptor that refers to a device.
+
+The request argument selects the control function to be performed and shall depend on the device being addressed.
+
+The arg argument represents additional information that is needed by this specific device to perform the requested function. The type of arg depends upon the particular control request, but it shall be either an integer or a pointer to a device-specific data structure.
+
+In order to define driver request one should use below macros and defines:
+
+    #define IOCPARM_MASK            0x3fff
+    #define IOCPARM_LEN(x)          (((x) >> 16) & IOCPARM_MASK)
+    #define IOCBASECMD(x)           ((x) & ~(IOCPARM_MASK << 16))
+    #define IOCGROUP(x)                     (((x) >> 8) & 0xff)
+
+
+    #define IOC_VOID                        0x00000000
+    #define IOC_OUT                         0x40000000
+    #define IOC_IN                          0x80000000
+    #define IOC_INOUT                       (IOC_IN | IOC_OUT)
+    #define IOC_DIRMASK                     0xe0000000
+
+
+    _IOC(inout,group,num,len)       ((unsigned long) (inout | ((len & IOCPARM_MASK) << 16) | ((group) << 8) | (num)))
+    _IO(g,n)                        _IOC(IOC_VOID, (g), (n), 0)
+    _IOV(g,n,t)                     _IOC(IOC_VOID, (g), (n), sizeof(t))
+    _IOR(g,n,t)                     _IOC(IOC_OUT, (g), (n), sizeof(t))
+    _IOW(g,n,t)                     _IOC(IOC_IN, (g), (n), sizeof(t))
+    _IOWR(g,n,t)                    _IOC(IOC_INOUT, (g), (n), sizeof(t))
```

Comment:
I think we shouldn't present there implementation details, which could potentially change.

```suggestion
 * _IOC(inout,group,num,len) - short description
 * _IOV(g,n) - short description
 * _IOR(g,n,t) 
 * _IOW(g,n,t) 
 * _IOWR(g,n,t) 
```

Above that I would not describe the exact bits, but suggest to use those macros and explain how to use its arguments. For example: One of the following macros shall be passed as _inout_ argument:

 * IOC_VOID - direction not specified
 * IOC_OUT - 
 ...

Group type shall be passed as _group_, typically character are used, for example 'S' for sockets, ... It has to be 8bit long, ...

And please describe this way the remaining arguments, adding that arguments for the rest of macros shall be passed by analogy.


Above that, some simple piece of code with an example could be very helpful. Of course it's only pattern, you don't have to use the same words etc.

Reply:
Please explain also what is `t` argument.

Reply:
Please say sth about using `ioctl_unpack()`, which is needed during writing an ioctl handler in a driver.